draft-irtf-cfrg-frost.md

```diff
@@ -224,12 +224,49 @@ for more details about each.
 Beyond the core dependencies, the protocol in this document depends on the
 following helper operations:
 
+- Nonce generation, {{dep-nonces}}
 - Schnorr signatures, {{dep-schnorr}};
 - Polynomial operations, {{dep-polynomial}};
 - Encoding operations, {{dep-encoding}};
 - Signature binding {{dep-binding-factor}}, group commitment {{dep-group-commit}}, and challenge computation {{dep-sig-challenge}}
 
-This sections describes these operations in more detail.
+These sections describes these operations in more detail.
+
+## Nonce generation {#dep-nonces}
+
+Straightforward nonce generation is simple:
+
+~~~
+  nonce_generate():
+
+  Outputs: Scalar nonce
+
+  def nonce_generate():
+    return G.RandomNonzeroScalar()
+~~~
+
+This fully relies on the good randomness of the random number generation powering `G.RandomNonZeroScalar()`.
+You may choose to generate randomness to combine with the secret key to hedge against a bad RNG, to create a domain-separated hash function from the ciphersuite hash function `H`:
+
+~~~
+  nonce_generate(secret_share):
+  
+  Inputs:
+  - secret_share, a Scalar in GF(p).
+
+  Outputs: Scalar nonce
+
+  def nonce_generate(secret_share):
+    k = G.RandomNonzeroScalar()
+
+    k_enc = G.SerializeScalar(k)
+    secret_share_enc = G.SerializeScalar(secret_share)
+    hash_input = contextString || "nonce" || pk_enc || k_enc
+    nonce_bytes = H(hash_input)
+
+    return G.DeserializeScalar(nonce_bytes)
```

Comment:
Since `DeserializeScalar` is fallible, I'd recommend we just do this:

```
def nonce_generate(random_bytes):
   hash_input = H4(random_bytes)
```

and then define `H4` _exactly_ as we do for H1 and H2, just with a different domain separator.

Reply:
So we can do that even if it's 'optional'? (Optional because this was not part of the security proof)

Reply:
Also, as this function doesn't _sample_ any randomness, I might call it `nonce_derive` (or `derive_nonce`).